File: composite/go/composite.go
package composite

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync/atomic"
)

type FileSystemNode interface {
	fmt.Stringer
	GetPath() string
	CountNumOfFiles() int32
	CountSizeOfFiles() int64
}

// ============================================================

type File struct {
	Path string
}

func (f *File) GetPath() string {
	return f.Path
}

func (f *File) CountNumOfFiles() int32 {
	return 1
}

func (f *File) CountSizeOfFiles() int64 {
	fileInfo, _ := os.Stat(f.Path)
	return fileInfo.Size()
}

func (f *File) String() string {
	return fmt.Sprintf("File{path='%s', fileNums=%d, fileSizes=%d}", f.Path, f.CountNumOfFiles(), f.CountSizeOfFiles())
}

// ============================================================

type Directory struct {
	Path      string
	ParentDir *Directory
	SubNodes  SubNodes
	fileNums  int32
	fileSizes int64
}

func NewDirectory(path string, parentDir *Directory) *Directory {
	return &Directory{
		Path:      path,
		ParentDir: parentDir,
		SubNodes:  SubNodes{make([]FileSystemNode, 0)},
	}
}

func (d *Directory) GetPath() string {
	return d.Path
}

func (d *Directory) CountNumOfFiles() int32 {
	return d.fileNums
}

func (d *Directory) CountSizeOfFiles() int64 {
	return d.fileSizes
}

func (d *Directory) incrCachedCountNumOfFiles(delta int32) {
	atomic.AddInt32(&d.fileNums, delta)
	if d.ParentDir != nil {
		d.ParentDir.incrCachedCountNumOfFiles(delta)
	}
}

func (d *Directory) decrCachedCountNumOfFiles(delta int32) {
	atomic.AddInt32(&d.fileNums, -delta)
	if d.ParentDir != nil {
		d.ParentDir.decrCachedCountNumOfFiles(delta)
	}
}

func (d *Directory) incrCachedCountSizeOfFiles(delta int64) {
	atomic.AddInt64(&d.fileSizes, delta)
	if d.ParentDir != nil {
		d.ParentDir.incrCachedCountSizeOfFiles(delta)
	}
}

func (d *Directory) decrCachedCountSizeOfFiles(delta int64) {
	atomic.AddInt64(&d.fileSizes, -delta)
	if d.ParentDir != nil {
		d.ParentDir.decrCachedCountSizeOfFiles(delta)
	}
}

func (d *Directory) AddSubNode(fsn FileSystemNode) {
	if d.SubNodes.Add(fsn) {
		d.incrCachedCountNumOfFiles(fsn.CountNumOfFiles())
		d.incrCachedCountSizeOfFiles(fsn.CountSizeOfFiles())
	}
}

func (d *Directory) RemoveSubNode(fsn FileSystemNode) {
	if d.SubNodes.Remove(fsn) {
		d.decrCachedCountNumOfFiles(fsn.CountNumOfFiles())
		d.decrCachedCountSizeOfFiles(fsn.CountSizeOfFiles())
	}
}

func (d *Directory) String() string {
	return fmt.Sprintf("Directory{path='%s', fileNums=%d, fileSizes=%d, subs=%s}",
		d.Path, d.fileNums, d.fileSizes, d.SubNodes.String())
}

type SubNodes struct {
	nodes []FileSystemNode
}

func (sub *SubNodes) IndexOf(fsn FileSystemNode) int {
	idx := -1
	fsnPath := strings.ToLower(fsn.GetPath())
	for i, node := range sub.nodes {
		if strings.ToLower(node.GetPath()) == fsnPath {
			idx = i
			break
		}
	}
	return idx
}

func (sub *SubNodes) Add(fsn FileSystemNode) bool {
	if sub.IndexOf(fsn) == -1 {
		sub.nodes = append(sub.nodes, fsn)
		return true
	}
	return false
}

func (sub *SubNodes) Remove(fsn FileSystemNode) bool {
	if idx := sub.IndexOf(fsn); idx != -1 {
		sub.nodes = append(sub.nodes[:idx], sub.nodes[idx+1:]...)
		return true
	}
	return false
}

func (sub *SubNodes) String() string {
	strs := make([]string, len(sub.nodes))
	for i, node := range sub.nodes {
		strs[i] = node.String()
	}
	return fmt.Sprintf("[%s]", strings.Join(strs, ", "))
}

// ============================================================

type Composite struct{}

func (c Composite) walk(parentDir *Directory, path string) {
	fileInfos, err := os.ReadDir(path)
	if err != nil {
		return
	}

	for _, fileInfo := range fileInfos {
		fullPath := filepath.Join(path, fileInfo.Name())
		if fileInfo.IsDir() {
			dir := NewDirectory(fullPath, parentDir)
			parentDir.AddSubNode(dir)
			c.walk(dir, fullPath)
		} else {
			ff := &File{fullPath}
			parentDir.AddSubNode(ff)
		}
	}
}

func (c Composite) BuildFileSystemTree(basePath string) FileSystemNode {
	fileInfo, err := os.Stat(basePath)
	if os.IsNotExist(err) {
		return nil
	}
	if fileInfo.IsDir() {
		dir := NewDirectory(basePath, nil)
		c.walk(dir, basePath)
		return dir
	}
	return &File{basePath}
}

File: composite/java/Composite.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.LongAdder;
import java.util.stream.Stream;

abstract class FileSystemNode {
    protected String path;

    public FileSystemNode(String path) {
        this.path = path;
    }

    public String getPath() {
        return path;
    }

    public void setPath(String path) {
        this.path = path;
    }

    public abstract int countNumOfFiles();

    public abstract long countSizeOfFiles();

    @Override
    public String toString() {
        return String.format("{path='%s', fileNums=%d, fileSizes=%d}", path, countNumOfFiles(), countSizeOfFiles());
    }
}

// ============================================================

/**
 * 文件
 */
class File extends FileSystemNode {

    public File(String path) {
        super(path);
    }

    @Override
    public int countNumOfFiles() {
        return 1;
    }

    @Override
    public long countSizeOfFiles() {
        return new java.io.File(path).length();
    }

    @Override
    public String toString() {
        return String.format("File%s", super.toString());
    }
}

// ============================================================

/**
 * 目录
 */
class Directory extends FileSystemNode {
    private final Directory parentDir;
    private final SubNodes subNodes = new SubNodes();

    private final LongAdder countNumAdder = new LongAdder();
    private final LongAdder countSizeAdder = new LongAdder();

    public Directory(String path) {
        this(path, null);
    }

    public Directory(String path, Directory parentDir) {
        super(path);
        this.parentDir = parentDir;
    }

    @Override
    public int countNumOfFiles() {
        return countNumAdder.intValue();
    }

    @Override
    public long countSizeOfFiles() {
        return countSizeAdder.longValue();
    }

    private void incrCachedCountNumOfFiles(int delta) {
        countNumAdder.add(delta);
        if (parentDir != null) {
            parentDir.incrCachedCountNumOfFiles(delta);
        }
    }

    private void decrCachedCountNumOfFiles(int delta) {
        countNumAdder.add(-delta);
        if (parentDir != null) {
            parentDir.decrCachedCountNumOfFiles(delta);
        }
    }

    private void incrCachedCountSizeOfFiles(long delta) {
        countSizeAdder.add(delta);
        if (parentDir != null) {
            parentDir.incrCachedCountSizeOfFiles(delta);
        }
    }

    private void decrCachedCountSizeOfFiles(long delta) {
        countSizeAdder.add(-delta);
        if (parentDir != null) {
            parentDir.decrCachedCountSizeOfFiles(delta);
        }
    }

    public void addSubNode(FileSystemNode fsn) {
        if (subNodes.add(fsn)) {
            incrCachedCountNumOfFiles(fsn.countNumOfFiles());
            incrCachedCountSizeOfFiles(fsn.countSizeOfFiles());
        }
    }

    public void removeSubNode(FileSystemNode fsn) {
        if (subNodes.remove(fsn)) {
            decrCachedCountNumOfFiles(fsn.countNumOfFiles());
            decrCachedCountSizeOfFiles(fsn.countSizeOfFiles());
        }
    }

    @Override
    public String toString() {
        return String.format("Directory{path='%s', fileNums=%d, fileSizes=%d, subs=%s}",
                path, countNumOfFiles(), countSizeOfFiles(), subNodes);
    }
}

/**
 * 目录下的子节点
 */
class SubNodes {
    // 注意这里是 FileSystemNode 列表，即 Directory 的抽象父类的列表（这样能组合 File 和 Directory）
    private final List<FileSystemNode> nodes = new ArrayList<>();

    public int indexOf(FileSystemNode fsn) {
        int size = nodes.size();
        boolean found = false;
        int i = 0;
        for (; i < size; i++) {
            if (nodes.get(i).getPath().equalsIgnoreCase(fsn.getPath())) {
                found = true;
                break;
            }
        }
        return found ? i : -1;
    }

    public boolean add(FileSystemNode fsn) {
        if (indexOf(fsn) == -1) {
            nodes.add(fsn);
            return true;
        }
        return false;
    }

    public boolean remove(FileSystemNode fsn) {
        int i;
        if ((i = indexOf(fsn)) != -1) {
            nodes.remove(i);
            return true;
        }
        return false;
    }

    @Override
    public String toString() {
        return nodes.toString();
    }
}

// ============================================================

public class Composite {
    private void walk(Directory parentDir, String path) {
        try (Stream<Path> stream = Files.list(Path.of(path))) {
            stream.forEach(p -> {
                java.io.File file = p.toFile();
                String fullPath = file.getAbsolutePath();
                if (file.isFile()) {
                    File ff = new File(fullPath);
                    parentDir.addSubNode(ff);
                } else {
                    Directory dir = new Directory(fullPath, parentDir);
                    parentDir.addSubNode(dir);
                    walk(dir, fullPath);
                }
            });
        } catch (IOException ignore) {}
    }

    @SuppressWarnings("ClassEscapesDefinedScope")
    public FileSystemNode buildFileSystemTree(String basePath) {
        java.io.File root = new java.io.File(basePath);
        if (!root.exists()) {
            return null;
        }
        if (root.isFile()) {
            return new File(basePath);
        }
        Directory directory = new Directory(basePath);
        walk(directory, basePath);
        directory.removeSubNode(new File("/Users/calvinit/Downloads/ll.txt"));
        return directory;
    }
}
